File: client-mvvm/app/src/main/java/ua/meugen/android/rxjavatests_client/viewmodel/rest/ModelApi.java
package ua.meugen.android.rxjavatests_client.viewmodel.rest;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import ua.meugen.android.rxjavatests_client.model.responses.DataResponse;

@Singleton
public class ModelApi {

    private final RestApi restApi;
    private Observable<DataResponse> dataObservable;
    private Disposable disposable;

    @Inject
    public ModelApi(final RestApi restApi) {
        this.restApi = restApi;
    }

    public Observable<DataResponse> dataWithDelay(
            final long delay, final int count) {
        if (disposable != null) {
            disposable.dispose();
            disposable = null;
        }
        BehaviorSubject<DataResponse> dataSubject
                = BehaviorSubject.create();
        disposable = restApi.dataWithDelay(delay, count)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(dataSubject::onNext, dataSubject::onError);
        dataObservable = dataSubject.hide();
        return dataObservable;
    }

    public Observable<DataResponse> getDataObservable() {
        return dataObservable;
    }
}

File: client-kotlin/app/src/main/java/ua/meugen/android/rxjavatests_client/presenter/rest/ModelApi.kt
package ua.meugen.android.rxjavatests_client.presenter.rest

import javax.inject.Inject
import javax.inject.Singleton

import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subjects.BehaviorSubject
import ua.meugen.android.rxjavatests_client.model.DataResponse

@Singleton
class ModelApi @Inject constructor(private val restApi: RestApi) {
    var dataObservable: Observable<DataResponse>? = null
        private set
    private var subscription: Subscription? = null

    fun dataWithDelay(
            delay: Long, count: Int): Observable<DataResponse> {
        if (subscription != null) {
            subscription!!.unsubscribe()
            subscription = null
        }
        val dataSubject = BehaviorSubject.create<DataResponse>()
        subscription = restApi.dataWithDelay(delay, count)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ dataSubject.onNext(it) }, { dataSubject.onError(it) })
        dataObservable = dataSubject.asObservable()
        return dataObservable!!
    }
}
